src/file_path.rs

```diff
@@ -30,20 +27,29 @@ impl PathExt for Path {
     }
 }
 
-fn decode_tail_path(tail: &str) -> Result<Cow<'_, str>, StatusCode> {
-    match percent_decode_str(tail.trim_start_matches('/')).decode_utf8() {
-        Ok(p) => Ok(p),
-        Err(err) => {
-            tracing::debug!("dir: failed to decode route={:?}: {:?}", tail, err);
-            Err(StatusCode::UNSUPPORTED_MEDIA_TYPE)
-        }
-    }
+#[cfg(unix)]
+fn path_from_bytes(bytes: &[u8]) -> PathBuf {
+    use std::ffi::OsStr;
+    use std::os::unix::ffi::OsStrExt;
+
+    OsStr::from_bytes(bytes).into()
+}
+
+#[cfg(windows)]
+fn path_from_bytes(bytes: &[u8]) -> PathBuf {
+    // This should really be OsStr::from_encoded_bytes_unchecked() but it’s
+    // unsafe. With this fallback non-Unicode file names will result in 404.
+    String::from_utf8_lossy(bytes).into_owned().into()
+}
```

Comment:
> As to Windows, I’m not sure that creating a file with a non-Unicode name is even possible. I did verify that the fallback employed there works on Linux (for properly encoded file names) but I didn’t perform any testing on Windows.

I also don't know how to create such a file but I'm fine with the safe approach of converting the invalid characters anyway.
Btw I just did a general functional testing in Windows.